### ISSUE TITLE
http://pi.hole domain not resolved when using PiHole in Docker

### ISSUE THREAD

**Issue by jaworek** [pi-hole/docker-pi-hole#553]
This is a... 

- [ ] Request for a new or modified feature
- [X] Issue trying to run the docker image
- [ ] Issue trying to build / test / develop the docker image

## Description
After fresh installation and setup of Docker Pihole on RaspberryPi 3 it is not possible to access web dashboard via http://pi.hole although from README it seems that it should be possible. Everything else seems to be working correctly and admin panel is accessible using IP Address of the device that it is setup on.

## Expected Behavior
Going to http://pi.hole should open web admin page of the Pihole project.

## Actual Behavior
![image](https://user-images.githubusercontent.com/22678765/71839013-81536d00-30ba-11ea-84e0-0b7569cc470c.png)

## Steps to Reproduce and debugging done
1. Download newest version of Raspbian Lite (https://www.raspberrypi.org/downloads/raspbian/)
2. Flash it on SD card with Etcher
3. Start and connect to RaspberryPi via SSH (it has static ip)
4. Install Docker using convenience script (https://docs.docker.com/v17.12/install/linux/docker-ce/debian/#install-docker-ce-1)
5. Install needed dependencies for docker-compose
6. Install docker compose
7. Run docker-compose.yml from this repository (https://github.com/pi-hole/docker-pi-hole/blob/master/docker-compose.yml)
8. Open browser on the computer connected to the same network and go to http://pi.hole. **FAILURE**
9. Try accessing web dashboard via \<ip-address-of-device\>/admin. **SUCCESS**
10. Check if ads are blocked on the network. **SUCCESS**

## Debug steps I have tried
- I've tried installing older versions of docker images (4.3.2, 4.3, 4.2).
- I've installed PiHole using bash script provided in the main repository on clean system and separate SD card. This approach works without an issue and allows access to dashboard via http://pi.hole
- I've tried switching "Interface listening behavior" in Settings/DNS to "Listen on all interfaces, permit all origins"
- I've tried setting VIRTUAL_HOST variable to 'pi.hole' and 'http://pi.hole', but it did not fix the problem

- [X] I have tried destroying my container instance, pulling the newest image version, and re-creating a new container
- [X] I have tried running the nearly stock `docker run` example in the readme (removing any customizations I added)
- [X] I have tried running without my volume data mounts to eliminate volumes as the cause **(I reflashed whole system and set everything from scratch)**
- [X] I have searched this repository for existing issues and pull requests that look similar 
https://github.com/pi-hole/docker-pi-hole/issues/523

## Context and extra information
- I have posted this question originally on reddit and got information that at least one other user had similar problem: https://www.reddit.com/r/pihole/comments/ekcoja/httppihole_domain_not_resolved_when_using_pihole/

## Your Environment 

* Docker Host Operating System and OS Version: Latest version of Raspbian Buster Lite
* Docker Version: 19.03.5, build 633a0ea
* Hardware architecture: ARMv7, RaspberryPi 3 with 32GB SD card

**Comment by nathang21** [comment]
I had the same issue, but “pi.hole.lan” and “pi.hole.local.lan” worked for
me.

**Comment by jaworek** [comment]
@nathang21 Did you specify VIRTUAL_HOST environment variable to set it up? I tried setting VIRTUAL_HOST to 'pi.hole' and 'http://pi.hole', but with no luck. I tried your suggestion, but it also didn't work.

**Comment by nathang21** [comment]
I didn't no.

Also if it helps you debug, the ".lan" part is set from on your router's
search domain. I am using a Google WiFi and that is the default local dns
domain, some routers have .lan or .local by default, or allow you to
specify a custom domain.
https://en.wikipedia.org/wiki/Search_domain

Here's where my config is at if it helps.
https://gitlab.com/nathang21/nate-and-xtina-home/blob/master/pi-hole/docker-compose.yml

**Comment by jaworek** [comment]
@nathang21 I've set domain name to 'domek' and tried to access web dashboard on 'http://pi.hole.local' and 'http://pi.hole.domek', but still no luck. I'll check out your setup to see if it can help me.

**Comment by randy993** [comment]
I'm currently having exactly the same problem. I even tried creating an extra .conf file (98-hostrecords-dns.conf) with content `address=/pi.hole/192.168.1.12` but it does not seem to work either.

**Comment by FlorentLM** [comment]
Same problem here. However, I happen to also have WireGuard installed on the same Pi, and whenever my devices are connected to WireGuard (no matter if they're actually in the local network or away), then http://pi.hole is resolved without issue.

What could be the problem?

**Comment by nathang21** [comment]
Sorry for the delay, finally home from travelling. I had a small typo
before, but I just confirmed, both of these are what work for me.
http://pihole.lan/admin (via local dns search domain from router, I use a
Google Wifi which automatically registers .lan)
http://pihole.local/admin (automatic via mDNS if router supports, which
Google WiFI does).

**Comment by HunterDG** [comment]
have you set your router or computer DNS to point to pi.hole?

**Comment by jaworek** [comment]
@HunterDG I don't think this is a problem with my router, as it blocks and resolves other domains correctly. Moreover, if I install PiHole without using Docker (bash script from main repository), everything seems to be working as intended.

**Comment by nathang21** [comment]
Do you have the "hostname" value set on your container? Here's how my
compose file ñools for reference.

pi-hole:    container_name: pi-hole    hostname: pi-hole
domainname: pihole.lan

**Comment by jaworek** [comment]
@nathang21 Setting these 2 options in my docker-compose.yml did not make any change. I tried many addresses, but none seem to work for me.

**Comment by HunterDG** [comment]
@[jaworek](https://github.com/jaworek) the only DNS server that can resolve "pi.hole" is pi-hole.  Thus you must manually set either your devices (that wish to use pi-hole) or your router's primary DNS to the IP address of your Docker host (your raspberry pi in this case).  Alternatively you can set your router's DHCP settings to hand out the pi's IP address as primary DNS server, but this requires disconnecting/reconnecting or `ipconfig /release && ipconfig /renew` (for windows) on client devices.  

Try the following from a test computer and see if both resolve successfully.  If only the second one resolves correctly, then your primary DNS is not pointing to pi-hole (pi's IP address)

```
nslookup pi.hole

nslookup pi.hole <pi's IP address>
```

The first command queries whatever DNS server your test device is pointing to.  The second command queries pi-hole (by your pi's IP address) directly.

obviously replace <pi's IP address> with pi's IP address

**Comment by jaworek** [comment]
@HunterDG Thanks for your advice about using `nslookup`. It seems that everything is working correctly.

I set up my DD-WRT router as it was advised in second point of this tutorial: https://discourse.pi-hole.net/t/how-do-i-configure-my-devices-to-use-pi-hole-as-their-dns-server/245

I ran nslookup command on my computer and this is the result:
![image](https://user-images.githubusercontent.com/22678765/73020725-8b999900-3e26-11ea-83a9-6ee531af71e6.png)

Here is my router configuration:
![image](https://user-images.githubusercontent.com/22678765/73020770-9f44ff80-3e26-11ea-8909-5ad94a1dfd86.png)
![image](https://user-images.githubusercontent.com/22678765/73021121-3ca03380-3e27-11ea-8e86-c64bced3bd0d.png)

From what I see on my dashboard, it seems that DNS server is setup properly as it detects other devices in the network and actively blocks queries that are on the list:
![image](https://user-images.githubusercontent.com/22678765/73020802-a835d100-3e26-11ea-97f9-83bda13f1b33.png)
![image](https://user-images.githubusercontent.com/22678765/73021088-25614600-3e27-11ea-9333-debeefc83743.png)

I'll also reiterate that this problem does not occur at all when I run PiHole directly on the device (installed with bash script).

**Comment by HunterDG** [comment]
few things:

- your pi-hole (the app) is resolving "pi.hole" (the domain) to 0.0.0.0 (nothing), this means pi-hole doesn't know its own IP address.  I imagine at pi-hole>settings>system you'll find:

Pi-hole IPv4 address: | 0.0.0.0
-- | --

I believe the documentation could be more clear about the `ServerIP` environment parameter - it's **not** required _**for pi-hole to block ads**_, but it **is** required _**to access pi-hole via the pi.hole url**_.  



| Docker Environment Var. | Description |
| ----------------------- | ----------- |
| `ServerIP: <Host's IP>`<br/> **Recommended** | **--net=host mode requires** Set to your server's LAN IP, used by web block modes **and lighttpd bind address**

add it under the environment section of docker-compose.yml like: 

```
environment:
    ServerIP:192.168.1.2
```

- That FAQ is **old**/not updated for current pi-hole conditional forwarding functionality

1. `dhcp-option=6...` is the **same function** as the Static DNS UI settings, so your current config is trying to hand out both 192.168.1.2 and 1.1.1.1.  This config is is **questionable\*** (see footnote)

2. Custom Upstream DNS>local router workaround for local DNS resolution has been replaced with proper conditional forwarding functionality.  Remove your custom upstream and just check Cloudflare (both) on the left side.

3. Conditional forwarding in your case:

- [x] Use Conditional Forwarding

IP of your router | Local domain name
-- | --
192.168.1.1|domek



### Questionable Config
Clearly your current config is blocking **some** things, but I don't know how DDWRT handles the priority with **both** option 6 and Static DNS configured since they are the same thing.  Check the DHCP-aquired DNS settings on one of your devices to see if you are getting both 192.168.1.2 and 1.1.1.1.  Make sure pi-hole IP is primary and 1.1.1.1 is secondary - This will probably be the case, but really no point in doing it **both** ways.  I'd suggest removing `dhcp-option=6...` and only use Static DNS UI settings: pi-hole for DNS 1 and both cloudflare addresses for DNS 2/3 (these will be used if DNS 1 is not accessible/pihole breaks/crashes).

**Comment by jaworek** [comment]
@HunterDG Thank you for your suggestions. I played around with setup of docker-compose and router settings. I managed to make it work when I run PiHole container with `network_mode: host`.  I decided to stick to Option 2 (dnsmasq) from official tutorial, as it seems to work best for me. On router I only left `1.1.1.1` in my static DNS settings.

This is my current docker-compose setup:
![image](https://user-images.githubusercontent.com/22678765/73205612-a7ac7b80-4141-11ea-8151-888b8c9b9727.png)

This is how my DNS is setup:
![image](https://user-images.githubusercontent.com/22678765/73205620-ac712f80-4141-11ea-8671-7795458564d2.png)

I didn't mention it, but I had conditional forwarding set before:
![image](https://user-images.githubusercontent.com/22678765/73205628-b004b680-4141-11ea-9f53-e1dda4b09945.png)

For now it will work for me, although I think it would be better to only have needed ports exposed. If my current setup is the only way to make 'http://pi.hole' to work or there is something wrong/missing in current README, it would be good to update it, to save confusion for new users.
I'll play around with it a little bit more when I have some time and I'll post updates if I find better way to make it work.

**Comment by HunterDG** [comment]
I agree, the documentation needs to be modified. Host network mode (and Net_ADMIN) are not required unless pi-hole is serving as a DHCP server.  If you want to be able to access "pi.hole" **without HOST network mode**, you **must make use of the ServerIP environment variable**. 

While your config is working, it's not the intended config.  Just like DHCP option 6 and "Static DNS" are **accomplishing the same thing**, so is conditional forwarding>router + Custom DNS>router.

You're currently sending ALL DNS requests from client device>pi-hole>router>Cloudflare.

Intended setup is client devices>pi-hole>Cloudflare + Conditional Forwarding>Router (ALL DNS requests are handled by pi-hole, which will query Cloudflare **directly**, and forward **only local** (.domek) requests to your router)

If you're using conditional forwarding **AND** Custom DNS, both pointing to your router, you're defeating the purpose of Conditional Forwarding.  Custom DNS>Router>external DNS was necessary to get local (*.domek) resolution working **before** Conditional Forwarding feature was introduced to pi-hole.  

You'll also want the following (from the official docker-compose) for pi-hole boot-up self-tests/blocklist downloads if you set your router to hand out pi-hole as primary static DNS via DHCP.  This allows pi-hole direct access to Cloudflare, without going through docker-host's/router's DNS). Note, this is different from your DNS1/2 environment variables.  
```
    dns:
      - 127.0.0.1
      - 1.1.1.1
```

If you want a configuration that _almost_ "just works", pull my [dev branch](https://github.com/HunterDG/Pi-Hole-Monitoring/tree/dev), **modify the .env file with your info**, bring the stack up, make sure your test device has pi-hole as it's primary DNS, flush it's DNS cache, and you will be able to browse to "pi.hole".  (nslookup pi.hole 192.168.1.2 will return 192.168.1.2, not 0.0.0.0

Unfortunately you'll have to go into pi.hole admin>DNS and press "save" for conditional forwarding to start working.  [see issue 557](https://github.com/pi-hole/docker-pi-hole/issues/557)

**Comment by jaworek** [comment]
@HunterDG  Great! Thank you very much! Everything seems to be working fine and I learned something alongside.

I'll leave this issue open, so anyone with similar problem can refer to it.

I have one sidenote regarding this setup. If I understand correctly, currently if my raspberry pi goes down for some reason my DNS resolution will stop working on my network, as the only static DNS points to pihole. Can I setup second static DNS without having to worry that some request will omit pihole when it is working correctly?

**Comment by jaworek** [comment]
I have one other downside to this setup and I'm not sure if it is supposed to be like that.
![image](https://user-images.githubusercontent.com/22678765/73577009-1e1be700-447c-11ea-862c-3bccd6635c81.png)
As you can see on the screenshot, number of queries has sharply risen and now most of them are made by my router. Did I miss something with my setup? Is there any way to avoid it?

**Comment by HunterDG** [comment]
Seems like your client devices are pointing to your router for primary DNS, instead of pi-hole.  Open network settings of a client and check the DNS servers it is using.  Pi-hole IP address should be first/Primary. 

If you changed DDWRT DHCP “Static DNS 1” to pi-hole IP address, you may need to disconnect/reconnect your client devices to pick up the change, or `ipconfig /release && ipconfig /renew`

**Comment by jaworek** [comment]
You were right that my router was set as primary DNS. I think the problem was that I left option "Local DNS" pointing to the router. I changed it to my PiHole address, but it did not change anything. I removed DHCP option 6 as you suggested before. I tried resetting my router to reset all connections, but it did not change output that I get on web dashboard.

My main concern with getting so much queries from router is that it might affect performance of my network and makes it harder to analyse what other devices are doing.
 
![image](https://user-images.githubusercontent.com/22678765/73611376-7fa99600-45e1-11ea-8354-7f26a4e46933.png)

**Comment by HunterDG** [comment]
Yeah - seems like you have a DNS resolution loop.  Wow - I've googled around and found a ton of problem posts with DDWRT and pi-hole.

**You may need to disconnect/reconnect your client devices, or `ipconfig /release && ipconfig /renew` to pick up the change (to Static DNS 1 - otherwise known as "Primary DNS")**

Also, your pi-hole "queries answered by" graph shows that pi-hole is still forwarding all requests to your router, as if you still have "Custom upstream DNS" checked and pointed to your router.  

Turn that off, check ONLY Cloudflare on the left side.  

Here's what I would do in DDWRT:
Gateway: 0.0.0.0 (not sure why you have it set to 192.168.1.255 - that's the network's broadcast address and is not valid)
Local DNS: 0.0.0.0 (I believe this setting is not used unless DDWRT is in bridge mode)
Static DNS 1: 192.168.1.2 (pi.hole)
Static DNS 2: 192.168.1.1 (DDWRT router - so clients can resolve .domek if pi-hole is offline)
Static DNS 3: 1.1.1.1 (Cloudflare - so clients can resolve internet if pi-hole is offline)

leave DNS2/3 0.0.0.0 if you don't want any requests to bypass pi-hole, but keep in mind your internet connectivity will break if pi-hole has a problem

ensure pi-hole is running with following DNS options (not DNS1/2 env variables)
```
    dns:
      - 127.0.0.1
      - 1.1.1.1
```

**Comment by NazarKosteckij** [comment]
After a year of pain whith editing `/etc/hosts` (can give you more details if need) each time I wanted to enter my pi-hole admin, I've found the problem and the most easiest solution. 

So the problem is that the pi.hole - is default hostname and there no arguments to override it. 
I just opened the logs and saw there following lines:
 ```
Added ENV to php:
                        "PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
                        "ServerIP" => "0.0.0.0",
                        "VIRTUAL_HOST" => "0.0.0.0",
```
So I added this ENV variables with proper values in to my runtime (docker container) and it helped. Now my pi-hole admin available from the internet.

**The solution:**
If you run docker image from `docker_run.sh` just edit `docker run` command it in this way: 
```
docker run -d \
    --hostname ${DOMAIN_NAME} \
    --name pihole \
    -p 53:53/tcp -p 53:53/udp \
    -p 80:80 \
    -e TZ="America/Chicago" \
    -e PROXY_LOCATION="${DOMAIN_NAME}" \
    -e ServerIP="${YOUR_IP}" \
    -e VIRTUAL_HOST="${DOMAIN_NAME}" \
    -v "$(pwd)/etc-pihole/:/etc/pihole/" \
    -v "$(pwd)/etc-dnsmasq.d/:/etc/dnsmasq.d/" \
    --dns=8.8.8.8 --dns=1.1.1.1 \
    --restart=unless-stopped \
    pihole/pihole:latest
```
**IMPORTANT:** replace all `${DOMAIN_NAME}` and `${YOUR_IP}` strings with your domain and ip

P.s. if someone like my post I'll create pull requst with this fix =)

**Comment by jaworek** [comment]
I played around with my setup on router and Raspberry Pi. I tried different configurations, but none seemed to fix the problem with queries from router being forwarded. Also, having set other DNS addresses seemed to have bypassed most of the blocking.
I went back to having my router using `dhcp-option=6`. Now logs are back to normal and I still can access it via `http://pi.hole`.
The only downside is now that when PiHole goes down it will not reroute me to other DNS. I don't think it will be a big problem, as it rarely happens and even if it does I can easily change settings on my router.

![image](https://user-images.githubusercontent.com/22678765/73783379-e40c5700-4793-11ea-9693-761ee4da84bb.png)
![image](https://user-images.githubusercontent.com/22678765/73783386-e7074780-4793-11ea-9085-15c90e76c8e0.png)
![image](https://user-images.githubusercontent.com/22678765/73783395-ec649200-4793-11ea-8f76-821578587447.png)
![image](https://user-images.githubusercontent.com/22678765/73783407-eff81900-4793-11ea-9224-f0d1e26f3b02.png)

**Comment by HunterDG** [comment]
@jaworek - you're still using Custom Upstream DNS in pi-hole.  Don't do that, select ONLY Cloudflare on the left side.  I've mentioned that about 5 times.

**Comment by jaworek** [comment]
I have had selected Cloudflare for my only DNS provider on my pihole. Still I had the issues that I have mentioned in my previous posts. When I switched back to `dhcp-option=6` and set Custom DNS as it was advised in the official guide. Most of my issues were resolved with the exception of not having DNS resolution if PiHole goes down, which is also mentioned in guide and for me this is the smallest problem, as it rarely happens (if at all) and I'm able to quickly resolve it.

**Comment by jaworek** [comment]
Here's the proof that it works as expected compared to previous one: 
![image](https://user-images.githubusercontent.com/22678765/73969859-6f185900-491c-11ea-9746-3475f6a729c8.png)

**Comment by jaworek** [comment]
This was my previous setup that has caused with suggested router setup problem with large number of queries. I want to clarify that at the time I had `dhcp-option=6` turned off, and DNS on my PiHole was pointing only to cloudflare. My static DNS was:
1. `192.168.1.2`
2. `192.168.1.1` (also tried without it)
3. `1.1.1.1` (also tried without it)
My local DNS on router was set to `0.0.0.0`.
I hope this clarifies any confusion about setup that I tried.
![image](https://user-images.githubusercontent.com/22678765/73970342-4e9cce80-491d-11ea-9b5c-a42c3043b0de.png)

**Comment by HunterDG** [comment]
Apologies, thanks for the clarification.

From your previous screenshot (with the spike in queries), note the highlighted areas:

![Screen Shot 2020-02-06 at 3 08 10 PM](https://user-images.githubusercontent.com/38483996/73979500-e5549580-48f3-11ea-8a6b-e251fb336e40.png)


The majority of queries are being answered by your router, meaning that they are being **correctly** conditionally forwarded through pi-hole (because they are queries to .domek domain)

I believe the spike was **accurate** and that some device on your network is querying excessively for no-thanks.invalid.**domek** and ton.smf1.twitter.com.**domek**

perhaps you could search the query history for he no-thanks.invalid.domek and ton.smf1.twitter.com.domek and see if you can figure out exactly WHAT device was sending those queries.

**Comment by skullydazed** [comment]
I'm running into this issue on newly setup pihole install today. I can access http://172.16.25.2/admin/ (the lan IP of my server) but no matter what I do it resolves pi.hole to 172.17.0.4 (the container IP for pihole.) 

```
$ nslookup pi.hole 172.16.25.2
Server:		172.16.25.2
Address:	172.16.25.2#53

Name:	pi.hole
Address: 172.17.0.4
```

I started the container with this command:

```
docker run -it --name pihole \
        -p 53:53/tcp \
        -p 53:53/udp \
        -p 67:67/udp \
        -p 80:80/tcp \
        -p 443:443/tcp \
        -e TZ=PST8PDT \
        -e WEBPASSWORD=******* \
        -e IPv6=false \
        -e ServerIP=172.16.25.2 \
        -v /etc/pihole:/etc/pihole \
        -v /etc/dnsmasq.d:/etc/dnsmasq.d \
        --dns=127.0.0.1 --dns=8.8.8.8 \
        --restart=unless-stopped \
        --hostname pi.hole \
        --detach=true \
        pihole/pihole:latest
```

**Comment by zaphar** [comment]
My pihole is also resolving the docker IP address even after setting the `ServerIP` environment variable. I notice that the /etc/hosts file has the docker address in there with the pi-hole address. Is it possible that dnsmasq is confusing that address with the pi.hole name?

**Comment by evanshortiss** [comment]
@zaphar @skullydazed add `--net=host` flag when starting pihole. I had this same issue and that resolved it.

I'm curious what committers such as @diginc think of this. If the `ServerIP` and `VIRTUAL_HOST` is set correctly, should the pihole respond with the value of `ServerIP` for queries for `VIRTUAL_HOST`?

For example `VIRTUAL_HOST=pi.hole` and `ServerIP=192.168.1.100`. In this case if I run pihole without `--net=host` it returns the internal IP of the container, but I'd have expected it to return `192.168.1.100`. 

This might simply be misunderstanding of the `ServerIP` variable, or is it a valid bug?

**Comment by skullydazed** [comment]
I missed the edit on @zaphar's comment, but removing the --hostname option also worked for me.

FTR I would personally prefer to avoid --net=host. Part of the reason I'm using docker is for isolation purposes.

**Comment by evanshortiss** [comment]
@skullydazed 100%, IMO that's a preferable solution for those of us using Docker 👍

**Comment by KerfuffleV2** [comment]
Pretty strange that this is still broken in the recommended run script more than 6 months later. I wasted a fair amount of time trying to figure out what was going wrong since I'd started with the official script.

For what it's worth, you don't actually have to remove `--hostname` — just set it to something other than `pi.hole`. I changed it to `pihole` and it correctly returns the `ServerIP` environment variable when looking up `pi.hole`

**Comment by beesarefriends** [comment]
Still having similar issues, this thread helped me get it working with the "ServerIP" docker environment variable

**Comment by PromoFaux** [comment]
Once we release image based on v6 (once we release v6!) the ServerIP variable will no longer be required.

**Comment by JulianHeuser** [comment]
This is still an issue for me, even on the latest docker image. Here's the result of nslookup:

(.177 is the ip of the pihole device)

`$ nslookup pi.hole
Server:		127.0.0.XX
Address:	127.0.0.XX#XX

** server can't find pi.hole: NXDOMAIN

$ nslookup pi.hole 192.168.1.177
Server:		192.168.1.177
Address:	192.168.1.177#53

Name:	pi.hole
Address: 192.168.1.177
Name:	pi.hole
Address: fd13:70ae:1a11:0:ba27:ebff:fe79:3d75
`

**Comment by yubiuser** [comment]
Pleas open a new issue.

**Comment by diego-velez** [comment]
I'm having the same issue on the latest docker image.

**Comment by rdwebdesign** [comment]
Please open a new issue.

**Comment by EnigmaG** [comment]
I found the answer, [link](https://discourse.pi-hole.net/t/dns-lookup-for-pi-hole-pointing-to-internal-ip-of-docker-container/77704/5)

Add to your compose file

```
FTLCONF_dns_reply_host_IPv4: '192.168.165.105'
FTLCONF_dns_reply_host_force4: true
```

with the correct IP
